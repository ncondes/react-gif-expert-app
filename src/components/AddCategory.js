import { useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ setCategories }) => {

  const [ inputValue, setInputValue ] = useState('');

  // handleInputChange
  const handleInputChange = e => setInputValue( e.target.value );
  

  // handleSubmit
  const handleSubmit = e => {
    e.preventDefault();

    if ( inputValue.trim().length > 2 ) {
      // para no tener que recibir la prop de category, podemos usar una funcion flecha para recibir el valor previo que se tenia de estado en categories.
      setCategories( prevCategories => [ inputValue, ...prevCategories ] );
      setInputValue('');
    }
  }

  return (
    <form onSubmit={ handleSubmit }>
      <input
        type='text'
        value={ inputValue }
        onChange={ handleInputChange }
      />
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
}

