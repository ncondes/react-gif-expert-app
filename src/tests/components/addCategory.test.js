import { AddCategory } from '../../components/AddCategory';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';

describe('Pruebas en el componente addCategory.js', () => {

  // se usa jest.fn() para conocer mas cosas de esta funcion como:
  // como fue llamada
  // fue llamada?
  // cuantas veces fue llamada?
  const setCategories = jest.fn();
  let wrapper;

  beforeEach( () => {
    jest.clearAllMocks();
    wrapper = shallow( <AddCategory
      setCategories={ setCategories }
    /> );
  })

  test('Debe de mostrarse correctamente', () => {

    expect( wrapper ).toMatchSnapshot();

  });

  test('Deberia de cambiar la caja de texto', () => {
    
    const input = wrapper.find('input');
    const value = 'Hola mundo'

    // para definir el evento de un input se hace {}
    input.simulate('change', {
      target: {
        value: value,
      }
    });

    expect( wrapper.find('p').text().trim() ).toBe( value )

  });
  
  test('No debe de postear la informacion con submit', () => {
    
    // simulamos el submit
    wrapper.find('form').simulate('submit', {
      preventDefault(){}
    });

    // saber si hemos llamado a la funcion 
    expect( setCategories ).not.toHaveBeenCalled();

  });

  test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
    
    const value = 'Hola mundo';
    // 1. simular el input change
    // estamos simulando un evento input change en el cual estamos poniendo como evento el value= 'Hola mundo'
    wrapper.find('input').simulate('change', { target: { value } });
    // console.log( wrapper.find('input').prop('value'), 'antes del submit' );


    // 2. simular el submit del formulario
    wrapper.find('form').simulate('submit', { preventDefault(){} });

    // 3. se debe haber llamado el set categories
    // con la siguiente linea se comprueba que se hizo el llamado a set categories
    expect( setCategories ).toHaveBeenCalled();
    // si se quisiera evaluar y asegurar de que la funcion setCategories se haya llamado con una funcion se har√≠a
    expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) ) 

    // 4. el value del input debe de estar vacio
    // console.log( wrapper.find('input').prop('value'), 'despues del submit' );
    expect( wrapper.find('input').prop('value') ).toBe('');


  });
   
})