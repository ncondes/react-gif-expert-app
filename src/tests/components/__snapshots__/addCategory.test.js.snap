// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente addCategory.js Debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
     
     
  </p>
  <input
    onChange={[Function]}
    placeholder="Example: Los Simpsons..."
    type="text"
    value=""
  />
</form>
`;
